test/Base_Tests/src/Semantic/Multi_Value_Convert_Spec.enso

```diff
@@ -0,0 +1,91 @@
+from Standard.Base import all
+from Standard.Test import all
+import Standard.Base.Errors.Common.Type_Error
+import Standard.Base.Errors.Common.No_Such_Method
+
+import project.Data.Complex.Complex
+
+type A
+    A value
+    a self = "a"
+type B
+    B value
+    b self = "b"
+type C
+    C value
+    c self = "c"
+
+A.from (that:Float) = A.A that
+B.from (that:Float) = B.B that
+C.from (that:Float) = C.C that
+
+type U
+    U value
+    u self = "u"
+type V
+    V v
+
+U.from (that:B) = U.U that.value
+V.from (that:C) = V.V that.value
+
+add_specs suite_builder =
+    suite_builder.group "Convert Multi Value" group_builder->
+
+        group_builder.specify "performs conversion to U&V from B and C" <|
+            abc = 3.14 : A&B&C
+            uv = abc : U&V
+
+            uv.u . should_equal "u"
+            uv.v . should_equal 3.14
+
+            Test.expect_panic No_Such_Method <| uv.a 
+            Test.expect_panic No_Such_Method <| uv.b 
+            Test.expect_panic No_Such_Method <| uv.c 
+            Test.expect_panic Type_Error (uv:A)
+            Test.expect_panic Type_Error (uv:B)
+            Test.expect_panic Type_Error (uv:C)
```

Comment:
I'm confused about this test.

Isn't a type-asserting expression `(uv:A)` supposed to work as the 'explicit' `instanceof`-like conversion? Then it probably should be able to extract the `A` type hidden from the multi-value `abc` upon the `uv = abc : U&V` conversion, no?

If that was a method call, e.g.:
```
f (a:A) = ...
...
f uv
```
then I agree this should yield a type error.

But here it is a 'type-asserting expression' which I thought according to our spec was supposed to act like `instanceof` and uncover any 'hidden' part of the type. What am I missing here?

Reply:
This test is great and it tests both method dispatch and type asserting expressions (although as noted above I'm not sure if the behaviour is correct?). 

But the test seems to be lacking any argument-type check tests. Can we add these? These would be similar to the expressions `uv:A` but instead we'd pass the `uv` value to a function with checked argument type:
```
f_a a:A = a.a
```
And see that if the type is 'hidden' then it cannot be uncovered in the argument type check.

Reply:
Argument checking test is here https://github.com/enso-org/enso/pull/11600/files#diff-fb0aa503a2a28b6e78764cf7ae6642ffdc2bfb775fa08e0ee7ce73b44fde4d45R97

Reply:
> it is a 'type-asserting expression' which I thought according to our spec was supposed to act like `instanceof` and uncover any 'hidden' part of the type

This test has been integrated in 1167b52f9aaeac5c9f48af9a59e16d69340735db together with documentation for "converting type check". The test checks the behavior exactly as documented.

Remember, we are [implementing variant 1](https://github.com/enso-org/enso/pull/11600#discussion_r1878591270). E.g. when we have to built new _intersection type_ by using conversions, then we start from scratch.

Reply:
So we are in disagreement over what variant 1 constitutes. I know you don't like my pseudocode but I used it as it allowed me to precisely describe what I mean, perhaps I could try using normal words but I'm sure I wouldn't be as precise.

In my variant 1 the explicit cast removes the 'unused' types from the visible type (`Set<T> resultTypes = Set.empty();`) but it actually still keeps them in the hidden part (`List<T> remainingTypes = allInputTypes.removeAll(resultTypes); return RuntimeType(resultTypes.toList(), remainingTypes);`. But your solution instead removes them altogether, 'clearing' the hidden part.

Are you sure that is right? How does that relate to the subclass metaphor? If a conversion is used, does that mean that we are losing the 'subclass' altogether and getting a new fresh class?

If we did `uv = abc:(A&B&C&U&V)` we'd have kept all the types, right?

Reply:
> what variant 1 constitutes

OK. Then I stop calling it variant 1.

> solution instead removes them altogether, 'clearing' the hidden part

Yes. When you have `Text.from (that:Integer)` and do `42:Text` it also removes `Integer` and creates completely new value unrelated to `42`.

> How does that relate to the subclass metaphor? If a conversion is used, does that mean that we are losing the 'subclass' altogether and getting a new fresh class?

Yes, that's the idea. Otherwise we would get the _power set_ of _intersection types_ behavior as I mentioned [here](https://github.com/enso-org/enso/pull/11600#discussion_r1878591270) and would rather avoid it.

> If we did `uv = abc:(A&B&C&U&V)` we'd have kept all the types, right?

Yes, `uv` would have all the types because _they were spelled_ in the code as `A&B&C&U&V`

Reply:
> have `Text.from (that:Integer)` and do `42:Text` it also removes `Integer` and creates completely new value unrelated to `42`.

Technically we could _return a multi value_ from each conversion representing both the original value as well the new one. In such a system `42:Text` would result in `42:Text&Integer:Text` multi value. Possible if useful for anything, but again it leads to the [power set problem](https://github.com/enso-org/enso/pull/11600#discussion_r1878591270).